clients/src/main/java/org/apache/kafka/common/network/KafkaChannel.java

```diff
@@ -72,16 +74,28 @@ public void prepare() throws IOException {
             transportLayer.handshake();
         if (transportLayer.ready() && !authenticator.complete())
             authenticator.authenticate();
+        if (transportLayer.ready() && authenticator.complete())
+            state = ChannelState.READY;
     }
 
     public void disconnect() {
         disconnected = true;
         transportLayer.disconnect();
     }
 
+    public void state(ChannelState state) {
+        this.state = state;
+    }
+
+    public ChannelState state() {
+        return this.state;
+    }
 
     public boolean finishConnect() throws IOException {
-        return transportLayer.finishConnect();
+        boolean connected = transportLayer.finishConnect();
+        if (connected)
+            state = ready() ? ChannelState.READY : ChannelState.AUTHENTICATE;
```

Comment:
Yes, the previous proposal was purely for SASL. But it made sense to do the warnings for both TLS and SASL since we can, so both TLS and SASL handshake failures result in the same warning with this PR.